src/ring_theory/dedekind_domain/adic_valuation.lean

```diff
@@ -331,4 +331,84 @@ instance adic_completion.has_lift_t : has_lift_t K (v.adic_completion K) :=
 /-- The ring of integers of `adic_completion`. -/
 def adic_completion_integers : valuation_subring (v.adic_completion K) := valued.v.valuation_subring
 
+instance : inhabited (adic_completion_integers K v) := ⟨0⟩
+
+variables (R K)
+
+lemma adic_completion.is_integer (x : v.adic_completion K) :
+  x ∈ v.adic_completion_integers K ↔ (valued.v x : ℤₘ₀) ≤ 1 :=
+iff.rfl
+
+section algebra_instances
+
+lemma adic_valued.has_uniform_continuous_const_smul' :
+  @has_uniform_continuous_const_smul R K v.adic_valued.to_uniform_space _ :=
+@has_uniform_continuous_const_smul_of_continuous_const_smul R K _ _ _
+    v.adic_valued.to_uniform_space _ _
+
+lemma adic_valued.has_uniform_continuous_const_smul :
+  @has_uniform_continuous_const_smul K K v.adic_valued.to_uniform_space _ :=
+@ring.has_uniform_continuous_const_smul K _ v.adic_valued.to_uniform_space _ _
+
+instance adic_completion.algebra' : algebra R (v.adic_completion K) :=
+@uniform_space.completion.algebra K _ v.adic_valued.to_uniform_space _ _ R _ _
+  (adic_valued.has_uniform_continuous_const_smul' R K v)
+
+@[simp] lemma coe_smul_adic_completion' (r : R) (x : v.adic_completion K) :
+  (↑(r • x) : v.adic_completion K) = r • (x : v.adic_completion K) :=
+rfl
+
+instance : algebra K (v.adic_completion K) :=
+@uniform_space.completion.algebra' K _ v.adic_valued.to_uniform_space _ _
+
+@[simp] lemma coe_smul_adic_completion (k : K) (x : v.adic_completion K) :
+  (↑(k • x) : v.adic_completion K) = k • (x : v.adic_completion K) :=
+rfl
+
+instance : is_scalar_tower R K (v.adic_completion K) :=
+@uniform_space.completion.is_scalar_tower R K K v.adic_valued.to_uniform_space _ _ _
+  (adic_valued.has_uniform_continuous_const_smul' R K v) _ _
+
+instance : algebra R (v.adic_completion_integers K) :=
+({ to_fun := λ r, ⟨coe $ algebra_map R K r, by simpa only [adic_completion.is_integer,
+    valued.valued_completion_apply] using v.valuation_le_one _⟩,
+  map_one' := by simpa only [ring_hom.map_one],
+  map_mul' := λ x y,
+  begin
+    ext,
+    simp_rw [ring_hom.map_mul, subring.coe_mul, subtype.coe_mk, uniform_space.completion.coe_mul],
+  end,
+  map_zero' := by simpa only [ring_hom.map_zero],
+  map_add' := λ x y,
+  begin
+    ext,
+    simp_rw [ring_hom.map_add, subring.coe_add, subtype.coe_mk, uniform_space.completion.coe_add],
+  end, } : R →+* v.adic_completion_integers K).to_algebra
+
+@[simp] lemma coe_smul_adic_completion_integers (r : R) (x : v.adic_completion_integers K) :
+  (↑(r • x) : v.adic_completion K) = r • (x : v.adic_completion K) :=
+by simpa only [algebra.smul_def]
```

Comment:
This should be true by `rfl`; the fact that its not suggests that the algebra instance above is not defined in the right way.

Reply:
The way to solve this problem is to use the `{}` constructor for `algebra` rather than the generally bad `ring_hom.to_algebra`.

Reply:
I have rewritten the algebra instance with the explicit constructor, but this still isn't true by rfl. @ocfnash, do you have any suggestions for this?

Reply:
If we want this to be true by `rfl` we can do the following (the two `sorry` should be easy)
```
instance : algebra R (v.adic_completion_integers K) :=
{ smul      := λ r x, ⟨r • (x : v.adic_completion K), sorry⟩,
  to_fun    := λ r, ⟨coe $ algebra_map R K r, by simpa only [adic_completion.is_integer,
    valued.valued_completion_apply] using v.valuation_le_one _⟩,
  map_one'  := by simp only [map_one]; refl,
  map_mul'  := λ x y,
  begin
    ext,
    simp_rw [ring_hom.map_mul, subring.coe_mul, subtype.coe_mk, uniform_space.completion.coe_mul],
  end,
  map_zero' := by simp only [map_zero]; refl,
  map_add'  := λ x y,
  begin
    ext,
    simp_rw [ring_hom.map_add, subring.coe_add, subtype.coe_mk, uniform_space.completion.coe_add],
  end,
  commutes' := λ r x, by rw mul_comm,
  smul_def' := λ r x, sorry }

@[simp] lemma coe_smul_adic_completion_integers (r : R) (x : v.adic_completion_integers K) :
  (↑(r • x) : v.adic_completion K) = r • (x : v.adic_completion K) :=
rfl
```
To be honest I don't know what is better.

Reply:
In topology/uniform_space/completion.lean we have
```
@[simp] lemma map_coe (hf : uniform_continuous f) (a : α) : (completion.map f) a = f a :=
cpkg.map_coe cpkg hf a
```

and that's not `rfl` and so I think this shouldn't be `rfl` either, but it is `@[simp]` so I think the `simpa` proof is fine.

Reply:
I filled in Riccardo's sorrys. Let me know if I should go back to the previous version.